lightning/src/blinded_path/message.rs

```diff
@@ -227,7 +227,10 @@ pub struct MessageForwardNode {
 
 /// TLVs to encode in an intermediate onion message packet's hop data. When provided in a blinded
 /// route, they are encoded into [`BlindedHop::encrypted_payload`].
+#[derive(Clone)]
 pub(crate) struct ForwardTlvs {
+	/// The padding data used to make all packets of a blinded path the same size
+	pub(crate) padding: Option<Padding>,
```

Comment:
Hmm... there was some earlier discussion around this approach and it was implemented here: https://github.com/lightningdevkit/rust-lightning/commit/b1436f8d8989198369a09df3f5b30851a0b5d3dc. The shortcoming was that you could use `WithPadding` to write an invalid TLV stream having two type `1` TLV records. Maybe we don't care so long as we always use it right?

Reply:
> Hmm, that design constraint assumes that we want to pad to make each hop equal (by looking at the max hop length), but I'm not sure that's what we want anyway. Doing so may leak something about the contents still - if we assume the last hop is the max, then we leak the type of context (okay, generally fine), but in the future may leak some info based on if some field is set or not. Instead, we could have a simple PADDING_TARGET_LENGTH constant and always pad to that length in both `ForwardTlvs::write` and `ReceiveTlvs::write` (with appropriate debug assertions that by default we never have hops longer than that).

My thoughts regarding padding was more in the context of dummy hops in order to hide which hop is the recipient, not necessarily to prevent leaking the intended use from a forwarder.

Regarding `PADDING_TARGET_LENGTH`, we couldn't use that for `MessageContext::Custom` since its size is arbitrary.

> This may not be appropriate for things that go in QR codes, but in those cases we may well not want to pad at all because we're super space constrained (and we'll want to just manually verify that we are always writing paths where forward hops are always the same length and receive hops are always the same length). But there it seems like we should `WriteableArgs` rather than `Writeable`?

Just so we have some concrete numbers, for an `Offer` using compact blinded paths, we use 10 bytes per forward hop and 28 bytes for the receive hop. For non-compact blinded paths, the forward hops use 35 bytes each. So we would lose much of the savings by padding each to 28 bytes in the compact case.

Regarding `WriteableArgs`, we don't have such a trait (yet) -- only `ReadableArgs`. For use here, the length would need to be piped through `construct_blinded_hops` to `encrypt_payload`, and we'd need to change or have a variation of `ChaChaPolyWriteAdapter` to implement (and wrap a) `WriteableArgs`.

Though, I'm not sure I quite understand what you mean by "manually verify ..." above. Could you give a more concrete example with a blinded path for an offer?

Idle thought: I recall some conversations about concatenating blinded paths. Is it possible to take a blinded path from an offer and concatenate dummy hops to the *end* of it? That would seem useful when requesting an invoice to help obscure who the final recipient is, assuming all hops were equally padded.

Reply:
> Wait, this PR is just about padding to make individual hops equal-sized, right? Anyone who sees a BP can see how many hops it is.

Correct. There's a yet-to-be-opened follow-up PR adding dummy hops.

> Mmm, yea, that would be a bit annoying code quantity. Another option would be a `WithPadding` wrapper that wraps a `Forward/ReceiveTlvs` and specifies the amount of padding we want. We could even drop the direct `Writeable` on `{Forward,Receive}Tlvs` at that point. I think my real complaint here is that the API ends up super surprising - a user might build a `ReceiveTlvs` object manually, add (or not) their own `Padding` and expect that to be used, but instead LDK silently replaces whatever the user told us to do with what LDK thinks it should do. I think LDK is probably right and should make that decision, but having it in the API is really weird.

I think that's about what the other branch was doing. So we can probably just do that again.

Note that the TLV structs for `BlindedMessagePath` are actually `pub(crate)` but the ones for `BlindedPaymentPath` are `pub`, unfortunately. Setting that aside, having the `padding` field didn't seem horrible given it is an actual TLV record defined in the spec. Whereas other serialization concerns tend to be something like the way an integer is encoded or whether a length is needed, which I agree shouldn't leak into the struct. But I'd rather not have the `padding` field visible in the public API, too.

> Not sure its related, if we want to hide the hop count we can just add dummy hops, this doesn't hide the hop count.

I was think in terms of an offer QR where adding dummy hops would be expensive. If the sender added them, then forwarding nodes wouldn't know where they are along the path. It wouldn't affect the sending node from trying to infer the destination based on the `NetworkGraph`, though.

Reply:
> We could even drop the direct `Writeable` on `{Forward,Receive}Tlvs` at that point.

Not sure if we could drop that because we need it to determine the serialized length without the padding. Maybe it would be implemented on those tupled with `Option<u64>` to determine whether or not to include padding?

Reply:
> Idle thought: I recall some conversations about concatenating blinded paths. Is it possible to take a blinded path from an offer and concatenate dummy hops to the end of it? That would seem useful when requesting an invoice to help obscure who the final recipient is, assuming all hops were equally padded.

I don't think the sender can add dummy hops, per the spec: "the recipient needs to fully validate each dummy hop's onion payload to detect tampering (and must ensure that these hops have been used and not truncated)." 

> Ah, right. I was discussing this a bit with @valentinewallace and we concluded there's maybe a reason to do it for onion messages (but not clear if you can concatenate blinded paths without the introduction point having helped build the first BP?), but certainly for payments we wouldn't as it reduces your information available on failures.

Yeah, for onion messages if the sender concatenates a blinded path from themselves to the recipient's intro node using `next_blinding_override`, then said intro node can't tell that they're the intro node. Adding dummy hops doesn't affect whether forwarding nodes can tell where they are along the path, IIUC.

Reply:
> > We could even drop the direct `Writeable` on `{Forward,Receive}Tlvs` at that point.
> 
> Not sure if we could drop that because we need it to determine the serialized length without the padding. Maybe it would be implemented on those tupled with `Option<u64>` to determine whether or not to include padding?

Could have a method that calculates the serialized length without implementing the `Writeable` trait, I suppose.

Reply:
> I don't think the sender can add dummy hops, per the spec: "the recipient needs to fully validate each dummy hop's onion payload to detect tampering (and must ensure that these hops have been used and not truncated)."

Do we do that? Not sure I quite understand what the spec is saying, though. How can the recipient "ensure that these hops have been used"? IIUC, dummy hops come after the recipient hop.

> Yeah, for onion messages if the sender concatenates a blinded path from themselves to the recipient's intro node using `next_blinding_override`, then said intro node can't tell that they're the intro node. Adding dummy hops doesn't affect whether forwarding nodes can tell where they are along the path, IIUC.

Could you explain how that last part is the case?

Reply:
> Do we do that? Not sure I quite understand what the spec is saying, though. How can the recipient "ensure that these hops have been used"? IIUC, dummy hops come after the recipient hop.

We don't add dummy hops atm, therefore we also don't support validation for them (since dummy hops are created by the recipient for themselves). 

I think the intention of the spec is for dummy hops to be processed the same as normal hops, to avoid timing attacks. So IIUC they're supposed to be encoded along the lines of how we do phantom nodes, where we process it as a forwarding hop but the next hop is ourselves, and so on until we get to the innermost onion layer that contains our actual recipient payload. This is gleaned from discussions by eclair/cln, though, I would have to look at the eclair codebase to see how they actually do it. 

> Could you explain how that last part is the case?

I'm not sure how dummy hops could affect how intermediate nodes view their position in the path. Could you explain why you think that *is* the case? lol

Reply:
Oh, right! The payload is always the same size (i.e., never shrinks), so a forwarder can't tell if the next hop is the recipient or not.